packages/tasit-action/src/TasitAct.js

```diff
@@ -0,0 +1,144 @@
+import "ethers/dist/shims.js";
+// Note: ethers SHOULD be imported from their main object
+// shims aren't injected with package import
+import { ethers } from "ethers";
+
+class Utils {
+  static isAddress = address => {
+    return typeof address === "string" && address.match(/^0x[0-9A-Fa-f]{40}$/);
+  };
+
+  static isABI = abi => {
+    return abi && Array.isArray(abi);
+  };
+
+  // https://github.com/ethers-io/ethers.js/blob/db383a3121bb8cf5c80c5488e853101d8c1df353/src.ts/utils/properties.ts#L20
+  static isEthersJsSigner = signer => {
+    return signer && signer._ethersType === "Signer";
+  };
+}
+
+class Subscription {
+  #txPromise;
+  #provider;
+
+  constructor(txPromise, provider) {
+    this.#txPromise = txPromise;
+    this.#provider = provider;
+  }
+
+  on = async (trigger, callback) => {
+    let blockHeight = 0;
+
+    if (trigger === "confirmation") blockHeight = 1;
+    else {
+      throw new Error(`Invalid subscription trigger, use: ["confirmation"]`);
+    }
+
+    if (!callback || typeof callback !== "function") {
+      throw new Error(`Cannot subscribe without a function`);
+    }
+
+    const tx = await this.#txPromise;
+
+    this.#provider.on(tx.hash, async receipt => {
+      const message = {
+        data: {
+          confirmations: receipt.confirmations,
+        },
+      };
+
+      try {
+        await callback(message);
+      } catch (error) {
+        throw new Error(`Callback function with error: ${error.message}`);
+      }
+    });
+  };
+
+  removeAllListeners = async () => {
+    const tx = await this.#txPromise;
+    this.#provider.removeAllListeners(tx.hash);
+  };
+}
+
+export class Contract {
+  address;
+  #provider;
+  #contract;
+
+  constructor(address, abi, wallet) {
+    if (!Utils.isAddress(address) || !Utils.isABI(abi))
+      throw new Error(`Cannot create a Contract without a address and ABI`);
+
+    this.#provider = this.#getDefaultProvider();
+
+    const signerOrProvider = wallet
```

Comment:
I wonder if we should make a word for this that handles both concepts

Reply:
This line is important enough that I wonder if we shouldn't do it with a ternary operator. It's not the most readable syntax in my opinion, and it might lower the chances of someone looking at the code understanding this distinction we make here

Reply:
At a minimum, maybe a comment saying if there's a wallet, then use that wallet with the provider connected. Otherwise, just use the provider directly. I know that a comment is not warranted for most if statements, but for some reason I think it's useful here

Reply:
What do you prefer?

```javascript
// If there's a wallet, connect it with provider otherwise uses provider directly (for read operations only)
const signerOrProvider = wallet
      ? wallet.connect(this.#provider)
      : this.#provider;

this.#contract = new ethers.Contract(address, abi, signerOrProvider);
```
or
```javascript
if (wallet) {
  this.#contract = new ethers.Contract(
    address,
    abi,
    wallet.connect(this.#provider)
  );
} else {
  this.#contract = new ethers.Contract(address, abi, this.#provider);
}
```

Reply:
@marcelomorgado In general I'm open to either. In this specific instance, I think I prefer the latter, because it leaves us open to moving the `wallet.connect(this.#provider)` to a line before the contract instantiation happens, and that lets us handle errors more explicitly.

Reply:
Actually after giving it more thought, that could be done with either version

Reply:
So I don't have a strong preference - up to you!

Reply:
Is can be resolved?

Reply:
Yep, we went with one of the options above (A), so it can be resolved